File: PerlInterviewQuestions/findIslandsBDS.pl
use strict;
use warnings;
use Data::Dumper;

sub displayList {
    my $L = shift();

    for (my $i = 0; $i < scalar(@$L); $i++) {
        for (my $j = 0; $j < scalar(@{$$L[$i]}); $j++) {
            print $$L[$i][$j] . " ";
        }
        print "\n";
    }
}

sub generateList {
    my @L = (
        [ 1, 1, 2, 3, 3 ],
        [ 1, 1, 2, 3, 2 ],
        [ 1, 1, 2, 2, 2 ],
        [ 3, 3, 2, 2, 2 ],
        [ 1, 3, 1, 1, 1 ],
        [ 1, 1, 3, 2, 2 ]
    );
    return @L;
}

sub BFS {
    my $L = shift();
    my $x = shift;
    my $y = shift;
    my $val = shift;
    my $visited = shift();

    my @A = ("$x,$y");
    push(@$visited, "$x,$y");
    my @queue = ("$x,$y");
    while (scalar(@queue) > 0) {
        my @next = ();
        my ($x, $y) = split(/,/, shift(@queue));
        for (my $i = $x - 1; $i < $x + 2; $i++) {
            for (my $j = $y - 1; $j < $y + 2; $j++) {
                if ($i >= 0 && $i <= $#$L && $j >= 0 && $j < scalar(@{$$L[0]})) {
                    my $ij = "$i,$j";
                    if ($$L[$i][$j] == $val && !grep (/^$ij$/, @$visited)) {
                        push(@A, $ij);
                        push(@$visited, $ij);
                        push(@next, $ij);
                    }
                }
            }
        }
        push(@queue, @next);
    }

    return @A;
}

my @L = generateList();
print(displayList(\@L));

my @visited = ();
my @B = ();

for (my $i = 0; $i <= $#L; $i++) {
    for (my $j = 0; $j < scalar(@{$L[$i]}); $j++) {
        my $ij = "$i,$j";
        if (!grep (/^$ij$/, @visited)) {
            my @A = BFS(\@L, $i, $j, $L[$i][$j], \@visited);
            push(@B, join(" | ", @A));
            @A = ();
        }
    }
}
$Data::Dumper::Indent = 1;
print Dumper \@B;

File: PythonInterviewQuestions/findIslandsBDS.py
def displayList(a):
    for i in range(len(a)):
        for j in range(len(a[i])):
            print(a[i][j], end=" ")
        print()


def generateMatrix():
    L = [
        [1, 1, 2, 3, 3],
        [1, 1, 2, 3, 2],
        [1, 1, 2, 2, 2],
        [3, 3, 2, 2, 2],
        [1, 3, 1, 1, 1],
        [1, 1, 3, 2, 2]
    ]
    return L


def BFS(L, x, y, val, visited):
    A = [(x, y)]
    visited.append((x, y))
    queue = [(x, y)]
    while len(queue) > 0:
        next = []
        x, y = queue.pop(0)
        for i in range(x - 1, x + 2):
            for j in range(y - 1, y + 2):
                if (0 <= i < len(L)) and (0 <= j < len(L[0])):
                    if L[i][j] == val and (i, j) not in visited:
                        A.append((i, j))
                        visited.append((i, j))
                        next.append((i, j))
        queue.extend(next)

    return A

def main():
    L = generateMatrix()
    displayList(L)
    visited, B = [], []
    for i in range(len(L)):
        for j in range(len(L[0])):
            if (i, j) not in visited:
                A = BFS(L, i, j, L[i][j], visited)
                B.append(A)

    for b in B:
        print(b)


if __name__ == '__main__':
    main()
